builder/vsphere/common/step_export.go

```diff
@@ -0,0 +1,295 @@
+//go:generate struct-markdown
+//go:generate mapstructure-to-hcl2 -type ExportConfig
+
+package common
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/sha512"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/hashicorp/packer/builder/vsphere/driver"
+	"github.com/hashicorp/packer/common"
+	"github.com/hashicorp/packer/helper/multistep"
+	"github.com/hashicorp/packer/packer"
+	"github.com/hashicorp/packer/template/interpolate"
+	"github.com/pkg/errors"
+	"github.com/vmware/govmomi/nfc"
+	"github.com/vmware/govmomi/vim25/soap"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+// You may optionally export an ovf from VSphere to the instance running Packer.
+//
+// Example usage:
+//
+// ```json
+// ...
+//   "vm_name": "example-ubuntu",
+// ...
+//   "export": {
+//     "force": true,
+//     "output_directory": "./output_vsphere"
+//   },
+// ```
+// The above configuration would create the following files:
+//
+// ```
+// ./output_vsphere/example-ubuntu-disk-0.vmdk
+// ./output_vsphere/example-ubuntu.ovf
+// ```
+type ExportConfig struct {
+	// name of the ovf. defaults to the name of the VM
+	Name string `mapstructure:"name"`
+	// overwrite ovf if it exists
+	Force bool `mapstructure:"force"`
+	// include iso and img image files that are attached to the VM
+	Images bool `mapstructure:"images"`
+	// generate manifest using SHA 1, 256, 512. use 0 (default) for no manifest
+	Sha       int          `mapstructure:"sha"`
+	OutputDir OutputConfig `mapstructure:",squash"`
+	// Advanced ovf export options. Options can include:
+	// * mac - MAC address is exported for all ethernet devices
+	// * uuid - UUID is exported for all virtual machines
+	// * extraconfig - all extra configuration options are exported for a virtual machine
+	// * nodevicesubtypes - resource subtypes for CD/DVD drives, floppy drives, and serial and parallel ports are not exported
+	//
+	// For example, adding the following export config option would output the mac addresses for all Ethernet devices in the ovf file:
+
+	// ```json
+	// ...
+	//   "export": {
+	//     "options": ["mac"]
+	//   },
+	// ```
+	Options []string `mapstructure:"options"`
+}
+
+var sha = map[int]func() hash.Hash{
```

Comment:
If you change this to a type string you can get rid of this package level var and update the `newHash` function to use a switch statement - see comment below.

Reply:
the benefit of the variable is that it is just a reference to the function. no hash objects have been created